cluster/topology/healthtracking_dynamic.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package topology
+
+import (
+	"github.com/uber/aresdb/utils"
+	"sync"
+	"time"
+)
+
+const (
+	unhealthyRetryPeriodSeconds = 10
+)
+
+type healthiness struct {
+	healthy             bool
+	lastUpdateTimestamp time.Time
+}
+
+type healthTrackingDynamicTopoImpl struct {
+	sync.RWMutex
+
+	dynamicTopology  Topology
+	hostsHealthiness map[Host]*healthiness
+	closed           bool
+}
+
+func NewHealthTrackingDynamicTopology(opts DynamicOptions) (HealthTrackingDynamicTopoloy, error) {
+	dynamicTopo, err := NewDynamicInitializer(opts).Init()
+	if err != nil {
+		return nil, err
+	}
+
+	topo := &healthTrackingDynamicTopoImpl{
+		dynamicTopology:  dynamicTopo,
+		hostsHealthiness: make(map[Host]*healthiness),
+	}
+
+	return topo, nil
+}
+
+func (ht *healthTrackingDynamicTopoImpl) Get() Map {
```

Comment:
what is the difference of this class/function with existing StaticTopology or function automatically filter out unhealth nodes?

Reply:
existing one depends on etcd as the source of truth for datanode healthiness. it depends on several things that will NOT affect whether queries can be served:
 1. network connection between datanode and etcd
 2. network connection between broker and etcd
Using HealthTrackingTopology, broker can track the healthiness itself.

E.g. consider a scenario where: 
  - all datanodes are alive
  - connections between datanode and etcd are lost
  - connections between broker and etcd alive
  - connections between broker and datanode are alive
If we use DynamicTopology with `includeUnhealthy = false`, we would not be able to serve any query.
with HealthTrackingTopology we can still serve queries.